File: python/extruded_polygon.py

import numpy

from icecube.MuonGun import Surface
from icecube.dataclasses import make_pair

def convex_hull(points):
    """Computes the convex hull of a set of 2D points.
 
    Input: an iterable sequence of (x, y) pairs representing the points.
    Output: a list of vertices of the convex hull in counter-clockwise order,
      starting from the vertex with the lexicographically smallest coordinates.
    Implements Andrew's monotone chain algorithm. O(n log n) complexity.
    
    Lifted from http://code.icecube.wisc.edu/svn/sandbox/ckopper/eventinjector/python/util/__init__.py
    """
 
    # convert to a list of tuples
    points = [(p[0],p[1]) for p in points]
    
    # Sort the points lexicographically (tuples are compared lexicographically).
    # Remove duplicates to detect the case we have just one unique point.
    points = sorted(set(points))
 
    # Boring case: no points or a single point, possibly repeated multiple times.
    if len(points) <= 1:
        return points
 
    # 2D cross product of OA and OB vectors, i.e. z-component of their 3D cross product.
    # Returns a positive value, if OAB makes a counter-clockwise turn,
    # negative for clockwise turn, and zero if the points are collinear.
    def cross(o, a, b):
        return (a[0] - o[0]) * (b[1] - o[1]) - (a[1] - o[1]) * (b[0] - o[0])
 
    # Build lower hull 
    lower = []
    for p in points:
        while len(lower) >= 2 and cross(lower[-2], lower[-1], p) <= 0:
            lower.pop()
        lower.append(p)
 
    # Build upper hull
    upper = []
    for p in reversed(points):
        while len(upper) >= 2 and cross(upper[-2], upper[-1], p) <= 0:
            upper.pop()
        upper.append(p)
 
    # Concatenation of the lower and upper hulls gives the convex hull.
    # Last point of each list is omitted because it is repeated at the beginning of the other list. 
    hull = lower[:-1] + upper[:-1]
    
    # convert into numpy array
    return numpy.array(hull)

def hull_to_normals(points):
    # append first point at the end to close the hull
    points = numpy.append(points, [points[0]], axis=0 )
    
    vecs = points[1:]-points[:-1]
    magn = numpy.sqrt(vecs[:,0]**2 + vecs[:,1]**2)
    
    normals = numpy.array([vecs[:,1]/magn, -vecs[:,0]/magn, numpy.zeros(magn.shape)]).T
    
    return normals

def hull_to_lengths(points):
    # append first point at the end to close the hull
    points = numpy.append(points, [points[0]], axis=0 )
    
    vecs = points[1:]-points[:-1]
    
    return numpy.sqrt(vecs[:,0]**2 + vecs[:,1]**2)

def signed_area(points):
    """Returns the signed area of a given simple (i.e. non-intersecting) polygon.
    Positive if points are sorted counter-clockwise.
    """

    # append first point at the end to close the hull
    points = numpy.append(points, [points[0]], axis=0 )
    
    return numpy.sum(points[:,0][:-1]*points[:,1][1:] - points[:,0][1:]*points[:,1][:-1])/2.

class ExtrudedPolygon(Surface):
    """
    A convex polygon in the x-y plane, extruded in the z direction
    """
    def __init__(self, xy_points, z_range):
        """
        :param xy_points: a list of x-y coordinate pairs. The convex hull of
                          these points will form the edge of the surface in the
                          x-y plane
        :param z_range: a pair giving the lower and upper boundary of the
                        surface in z.
        """
        super(ExtrudedPolygon, self).__init__()
        assert len(xy_points) >= 3, "Need at least 3 points to form a closed polygon"
        
        hull = convex_hull(xy_points)
        
        # hull points, in counterclockwise order
        self._x = hull
        # next neighbor in the hull
        self._nx = numpy.roll(hull, -1, axis=0)
        # vector connecting each pair of points in the hull
        self._dx = self._nx - self._x
        
        self._z_range = z_range
        self.length = z_range[1] - z_range[0]
        self._side_lengths = hull_to_lengths(hull)
        
        side_normals = hull_to_normals(hull)
        side_areas = self._side_lengths*self.length
        cap_area = [signed_area(hull)]*2
        cap_normals = numpy.array([[0., 0., 1.], [0., 0., -1.]])
        
        self._areas = numpy.concatenate((side_areas, cap_area))
        self._normals = numpy.concatenate((side_normals, cap_normals))
        assert self._areas.size == self._normals.shape[0]
    
    def expand(self, padding):
        """
        Expand the x-y footprint by moving each edge out by a distance *padding*.
        """
        # A convex polygon can be offset by moving each vertex parallel to the
        # edges by a distance that is inversely proportional to the sine of the
        # counterclockwise angle between the edges that meet at each vertex.
        # This breaks down for edges that are [anti]parallel or, but neither
        # case should occur for maximally simplified polygons.
        
        # normalized vector connecting each vertex to the next one
        d = self._dx/self._side_lengths[:,None]
        # and the one connecting the previous vertex
        prev_d = numpy.roll(d, 1, axis=0)
        # sine of the inner angle of each vertex
        det = prev_d[:,0]*d[:,1] - prev_d[:,1]*d[:,0]
        assert (det != 0.).all(), "Edges can't be [anti]parallel"
        points = self._x + (padding/det[:,None])*(prev_d - d)
        
        z_range = [self._z_range[0]-padding, self._z_range[1]+padding]
        
        return type(self)(points, z_range)
    
    @classmethod
    def from_I3Geometry(cls, i3geo, padding=0):
        from collections import defaultdict
        strings = defaultdict(list)
        for omkey, omgeo in i3geo.omgeo:
            if omgeo.omtype != omgeo.IceTop:
                strings[omkey.string].append(list(omgeo.position))
        mean_xy = [numpy.mean(positions, axis=0)[0:2] for positions in strings.values()]
        zmax = max(max(p[2] for p in positions) for positions in strings.values())
        zmin = min(min(p[2] for p in positions) for positions in strings.values())
        
        self = cls(mean_xy, [zmin, zmax])
        if padding != 0:
            return self.expand(padding)
        else:
            return self
    
    @classmethod
    def from_file(cls, fname, padding=0):
        from icecube import icetray, dataio, dataclasses
        f = dataio.I3File(fname)
        fr = f.pop_frame(icetray.I3Frame.Geometry)
        f.close()
        return cls.from_I3Geometry(fr['I3Geometry'], padding)
    
    def area(self, dir):
        """
        Return projected area in the given direction
        
        :param dir: an I3Direction
        """
        # inner product with component normals
        inner = numpy.dot(self._normals, numpy.asarray((dir.x, dir.y, dir.z)))
        # only surfaces that face the requested direction count towards the area
        mask = inner < 0
        return -(inner*self._areas[:,None]*mask).sum(axis=0)
    
    def _point_in_hull(self, point):
        """
        Test whether point is inside the 2D hull by ray casting
        """
        x, y = point[0:2]
        # Find segments whose y range spans the current point
        mask = ((self._x[:,1] > y)&(self._nx[:,1] <= y))|((self._x[:,1] <= y)&(self._nx[:,1] > y))
        # Count crossings to the right of the current point
        xc = self._x[:,0] + (y-self._x[:,1])*self._dx[:,0]/self._dx[:,1]
        crossings = (x < xc[mask]).sum()
        inside = (crossings % 2) == 1
        
        return inside
    
    def _distance_to_hull(self, point, vec):
        """
        Calculate the most extreme displacements from x,y along dx,dy to points
        on the 2D hull
        """
        # calculate the distance along the ray to each line segment
        x, y = (self._x - point[:2]).T
        dx, dy = self._dx.T
        dirx, diry = vec[0:2]
        
        assert dirx+diry != 0, "Direction vector may not have zero length"
        
        # proportional distance along edge to intersection point
        # NB: if diry/dirx == dy/dx, the ray is parallel to the line segment
        nonparallel = diry*dx != dirx*dy
        alpha = numpy.where(nonparallel, (dirx*y - diry*x)/(diry*dx - dirx*dy), numpy.nan) 
        # check whether the intersection is actually in the segment
        mask = (alpha >= 0)&(alpha < 1)
        
        # distance along ray to intersection point
        if dirx != 0:
            beta = ((x + alpha*dx)/dirx)[mask]
        else:
            beta = ((y + alpha*dy)/diry)[mask]
        
        if beta.size == 0:
            return (numpy.nan,)*2
        else:
            return (numpy.nanmin(beta), numpy.nanmax(beta))
    
    def _distance_to_cap(self, point, dir, cap_z):
        d = (point[2]-cap_z)/dir[2]
        if self._point_in_hull(point + d*dir):
            return d
        else:
            return numpy.nan
    
    def _distance_to_caps(self, point, dir):
        return sorted((self._distance_to_cap(point, dir, cap_z) for cap_z in self._z_range))
    
    def GetIntersection(self, pos, dir):
        point = numpy.array((pos.x, pos.y, pos.z))
        vec = numpy.array((dir.x, dir.y, dir.z))
        
        # perfectly vertical track: only check intersections with caps
        if abs(dir.z) == 1.:
            return make_pair(*self._distance_to_caps(point, vec))
        # perfectly horizontal track: only check intersections with sides
        elif dir.z == 0.:
            return make_pair(*self._distance_to_hull(point, vec))
        # general case: both rho and z components nonzero
        else:
            sin_zenith = numpy.sqrt(1.-dir.z**2)
            sides = numpy.array(self._distance_to_hull(point, vec))/sin_zenith
            caps = self._distance_to_caps(point, vec)
            intersections = numpy.concatenate((sides, caps))
            return make_pair(*(numpy.nanmin(intersections), numpy.nanmax(intersections)))

File: private/MuonGun/ExtrudedPolygon.cxx
/** $Id$
 * @file
 * @author Jakob van Santen <jakob.van.santen@desy.de>
 *
 * $Revision$
 * $Date$
 */

#include <MuonGun/ExtrudedPolygon.h>
#include <dataclasses/I3Position.h>
#include <dataclasses/I3Direction.h>
#include <phys-services/I3RandomService.h>

#include <boost/foreach.hpp>
#include <cmath>

namespace {

typedef simclasses::polygon::vec2 vec2;

/// A counterclockwise curve is the basic building block of a convex hull
class ccw_curve : public std::vector<vec2> {
public:
	// Add a point to the curve
	void operator()(const vec2 &p)
	{
		// Remove points until the curve will be counterclockwise
		while (size() >= 2 && !ccw((*this)[size()-2], (*this)[size()-1], p))
			pop_back();
		push_back(p);
	}
private:
	static bool
	ccw(const vec2 &o, const vec2 &a, const vec2 &b)
	{
		// 2D cross product of OA and OB vectors, i.e. z-component of their 3D cross product.
		// positive, if OAB makes a counter-clockwise turn,
		// negative for clockwise turn, and zero if the points are collinear.
		return (a.x - o.x) * (b.y - o.y) - (a.y - o.y) * (b.x - o.x) > 0;
	}
};

/// @brief Computes the convex hull of a set of 2D points.
/// Input: an iterable sequence of (x, y) pairs representing the points.
/// Output: a list of vertices of the convex hull in counter-clockwise order,
///   starting from the vertex with the lexicographically smallest coordinates.
/// Implements Andrew's monotone chain algorithm. O(n log n) complexity.
/// 
/// Lifted from http://code.icecube.wisc.edu/svn/sandbox/ckopper/eventinjector/python/util/__init__.py
std::vector<vec2>
convex_hull(const std::vector<I3Position> &positions)
{
	std::vector<vec2> hull;
	
	// Build a set of unique points, sorted lexicographically
	std::set<vec2> points;
	std::transform(positions.begin(), positions.end(),
	    std::inserter(points, points.end()), vec2::from_I3Position);
	
	// Boring case: 1 point (perhaps repeated)
	if (points.size() <= 1) {
		std::copy(points.begin(), points.end(), std::back_inserter(hull));
		return hull;
	}
	
	// Build lower and upper hulls
	std::vector<vec2> lower = std::for_each(points.begin(), points.end(), ccw_curve());
	std::vector<vec2> upper = std::for_each(points.rbegin(), points.rend(), ccw_curve());
	
	// Concatenation of the lower and upper hulls gives the convex hull.
	// Last point of each list is omitted because it is repeated at the
	// beginning of the other list.
	std::copy(lower.begin(), lower.end()-1, std::back_inserter(hull));
	std::copy(upper.begin(), upper.end()-1, std::back_inserter(hull));
	
	return hull;
}


/// @brief Expand the x-y footprint by moving each edge out by a distance
/// *padding*.
/// 
/// A convex polygon can be offset by moving each vertex parallel to the
/// edges by a distance that is inversely proportional to the sine of the
/// counterclockwise angle between the edges that meet at each vertex.
/// This breaks down for edges that are [anti]parallel or, but neither
/// case should occur for maximally simplified polygons.
std::vector<vec2>
expand_polygon(const std::vector<vec2> &hull, double padding)
{
	std::vector<vec2> points;
	for (std::vector<vec2>::const_iterator p = hull.begin(); p != hull.end(); p++) {
		std::vector<vec2>::const_iterator next = boost::next(p);
		if (next == hull.end())
			next = hull.begin();
		std::vector<vec2>::const_iterator prev = boost::prior(
		    p == hull.begin() ? hull.end() : p);
		// normalized vector connecting this vertex to the next one
		vec2 d = vec2::normalized(next->x-p->x, next->y-p->y);
		// and the previous vertex to this one
		vec2 prev_d = vec2::normalized(p->x-prev->x, p->y-prev->y);
		// sine of the inner angle between the segments that meet here
		double det = prev_d.x*d.y - prev_d.y*d.x;
		if (det == 0.)
			log_fatal("Edges can't be [anti]parallel");
		vec2 outwards(prev_d.x-d.x, prev_d.y-d.y);
		points.push_back(vec2(p->x + outwards.x*padding/det, p->y + outwards.y*padding/det));
	}
	
	return points;
}

std::pair<double, double>
z_range(const std::vector<I3Position> &positions)
{
	std::pair<double, double> range(std::numeric_limits<double>::infinity(), -std::numeric_limits<double>::infinity());
	BOOST_FOREACH(const I3Position &pos, positions) {
		if (pos.GetZ() < range.first)
			range.first = pos.GetZ();
		if (pos.GetZ() > range.second)
			range.second = pos.GetZ();
	}
	
	return range;
}

std::pair<double, double>
make_ordered_pair(double a, double b)
{
	if (std::isnan(a) || (a > b))
		return std::make_pair(b, a);
	else
		return std::make_pair(a, b);
}

}

namespace simclasses {

namespace polygon {

vec2::vec2(double xi, double yi) : x(xi), y(yi)
{}

template <typename Archive>
void vec2::serialize(Archive &ar, unsigned version)
{
	if (version > 0)
		log_fatal_stream("Version "<<version<<" is from the future");
	
	ar & make_nvp("X", x);
	ar & make_nvp("Y", y);
}

vec2
vec2::from_I3Position(const I3Position &p)
{
	return vec2(p.GetX(), p.GetY());
}

vec2
vec2::normalized(double xi, double yi)
{
	double l = hypot(xi, yi);
	return vec2(xi/l, yi/l);
}

bool
operator<(const vec2 &a, const vec2 &b)
{
	if (a.x < b.x)
		return true;
	else if (a.x > b.x)
		return false;
	else if (a.y < b.y)
		return true;
	else
		return false;
}

side::side(const vec2 &p, const vec2 &np) : origin(p),
    vector(np.x-p.x, np.y-p.y), length(hypot(vector.x, vector.y)),
	normal(vector.y/length, -vector.x/length, 0.)
{}

}

template <typename Base>
ExtrudedPolygonBase<Base>::ExtrudedPolygonBase(const std::vector<I3Position> &points, double padding)
{
	using simclasses::polygon::vec2;
	
	std::pair<double, double> zrange = z_range(points);
	std::vector<vec2> hull = convex_hull(points);
	if (padding != 0) {
		hull = expand_polygon(hull, padding);
		zrange.first -= padding;
		zrange.second += padding;
	}
	initWithHull(hull, zrange);
}

template <typename Base>
void
ExtrudedPolygonBase<Base>::initWithHull(const std::vector<polygon::vec2> &hull,
    const std::pair<double, double> &z_range)
{
	using simclasses::polygon::vec2;
	using simclasses::polygon::side;
	
	z_range_ = z_range;
	cap_area_ = 0;
	sides_.clear();
	for (std::vector<vec2>::const_iterator p = hull.begin(); p != hull.end(); p++) {
		std::vector<vec2>::const_iterator np = boost::next(p);
		if (np == hull.end())
			np = hull.begin();
		
		sides_.push_back(side(*p, *np));
		
		// area of a simple polygon in the x-y plane
		cap_area_ += (p->x*np->y - np->x*p->y);
	}
	cap_area_ /= 2.;
}

template <typename Base>
std::vector<double>
ExtrudedPolygonBase<Base>::GetX() const
{
	std::vector<double> x;
	BOOST_FOREACH(const polygon::side &sidey, sides_)
		x.push_back(sidey.origin.x);
	return x;
}

template <typename Base>
std::vector<double>
ExtrudedPolygonBase<Base>::GetY() const
{
	std::vector<double> y;
	BOOST_FOREACH(const polygon::side &sidey, sides_)
		y.push_back(sidey.origin.y);
	return y;
}

template <typename Base>
std::vector<double>
ExtrudedPolygonBase<Base>::GetZ() const
{
	std::vector<double> z;
	z.push_back(z_range_.first);
	z.push_back(z_range_.second);
	
	return z;
}

/// Calculate the most extreme displacements to points on the 2D hull
template <typename Base>
std::pair<double, double>
ExtrudedPolygonBase<Base>::GetDistanceToHull(const I3Position &pos, const I3Direction &dir) const
{
	std::pair<double, double> offsets = Surface::no_intersection();
	
	if (dir.GetX() + dir.GetY() == 0.)
		log_fatal("Direction must have a horizontal component");
	
	BOOST_FOREACH(const polygon::side &sidey, sides_) {
		// Components of the vector connecting the test point to the
		// origin of the line segment;
		double x = sidey.origin.x - pos.GetX();
		double y = sidey.origin.y - pos.GetY();
		
		// Proportional distance along the line segment to the
		// intersection point
		double alpha = (dir.GetX()*y - dir.GetY()*x)
		    / (dir.GetY()*sidey.vector.x - dir.GetX()*sidey.vector.y);
		
		// Is there an intersection?
		if ((alpha >= 0.) && (alpha < 1.)) {
			// Distance along ray to intersection point
			double beta = dir.GetX() != 0. ?
			    (x + alpha*sidey.vector.x)/dir.GetX() :
			    (y + alpha*sidey.vector.y)/dir.GetY();
			// NB: reversed comparison is equivalent to
			// (isnan(offsets.first) || beta < offsets.first)
			if (!(beta >= offsets.first))
				offsets.first = beta;
			if (!(beta <= offsets.second))
				offsets.second = beta;
		}
	}
	double sin_theta = std::sqrt(1.-dir.GetZ()*dir.GetZ());
	offsets.first /= sin_theta;
	offsets.second /= sin_theta;
	
	return offsets;
}

/// Test whether point is inside the 2D hull by ray casting
template <typename Base>
bool
ExtrudedPolygonBase<Base>::PointInHull(const I3Position &pos) const
{
	int crossings = 0;
	for (std::vector<polygon::side>::const_iterator p = sides_.begin(); p != sides_.end(); p++) {
		std::vector<polygon::side>::const_iterator np = boost::next(p);
		if (np == sides_.end())
			np = sides_.begin();
		
		// only consider segments whose y range spans the current point
		if (!(((p->origin.y > pos.GetY()) && (np->origin.y <= pos.GetY())) ||
		      ((p->origin.y <= pos.GetY()) && (np->origin.y > pos.GetY()))))
			continue;
		// find the intersection of the segment with a horizontal line
		double xc = p->origin.x + (pos.GetY() - p->origin.y)*p->vector.x/p->vector.y;
		// is the crossing to the right of the test point?
		if (pos.GetX() < xc)
			crossings++;
	}
	
	// the point is inside iff the ray crosses an odd number of times
	return (crossings % 2) == 1;
}

template <typename Base>
double
ExtrudedPolygonBase<Base>::GetDistanceToCap(const I3Position &p, const I3Direction &dir, double cap_z) const
{
	return (cap_z-p.GetZ())/dir.GetZ();
}

template <typename Base>
std::pair<double, double>
ExtrudedPolygonBase<Base>::GetDistanceToCaps(const I3Position &p, const I3Direction &dir) const
{
	return make_ordered_pair(GetDistanceToCap(p, dir, z_range_.first), GetDistanceToCap(p, dir, z_range_.second));
}

template <typename Base>
std::pair<double, double>
ExtrudedPolygonBase<Base>::GetIntersection(const I3Position &p, const I3Direction &dir) const
{
	if (std::abs(dir.GetZ()) == 1.) {
		// perfectly vertical track: only check intersections with caps
		if (!PointInHull(p))
			return Surface::no_intersection();
		else
			return GetDistanceToCaps(p, dir);
	} else if (dir.GetZ() == 0.) {
		// perfectly horizontal track: only check intersections with sides
		if (p.GetZ() < z_range_.first || p.GetZ() > z_range_.second)
			return Surface::no_intersection();
		else
			return GetDistanceToHull(p, dir);
	} else {
		// general case: both rho and z components nonzero
		std::pair<double, double> sides = GetDistanceToHull(p, dir);
		std::pair<double, double> caps = GetDistanceToCaps(p, dir);
		
		if (caps.first >= sides.second || caps.second <= sides.first)
			return Surface::no_intersection();
		else {
			return std::make_pair(std::max(sides.first, caps.first), std::min(sides.second, caps.second));
		}
	}
}

template <typename Base>
double
ExtrudedPolygonBase<Base>::GetArea(const I3Direction &dir) const
{
	double area = 0;
	BOOST_FOREACH(const polygon::side &sidey, sides_) {
		double inner = sidey.normal*dir;
		if (inner < 0)
			area += -inner*sidey.length;
	}
	area *= (z_range_.second-z_range_.first);
	area += std::abs(dir.GetZ())*cap_area_;
	
	return area;
}

template <typename Base>
double
ExtrudedPolygonBase<Base>::GetMaximumArea() const
{
	double side_area = 0;
	BOOST_FOREACH(const polygon::side &sidey, sides_) {
		side_area += sidey.length;
	}
	// the largest possible projected area occurs for a flat square
	side_area *= (z_range_.second-z_range_.first)/2.;
	double ct_max = cos(atan(side_area/cap_area_));
	
	return cap_area_*fabs(ct_max) + side_area*sqrt(1.-ct_max*ct_max);
}

template <typename Base>
double
ExtrudedPolygonBase<Base>::GetAverageSideArea() const
{
	// the projected area of a plane, averaged over a 2\pi rotation that
	// passes through the normal, is
	// A*\int_0^\pi \Theta(\sin\alpha)\sin\alpha d\alpha / 2\pi = A/\pi
	double area = 0;
	BOOST_FOREACH(const polygon::side &sidey, sides_)
		area += sidey.length;
	area *= GetLength()/M_PI;
	
	return area;
}

template <typename Base>
I3Position
ExtrudedPolygonBase<Base>::SampleImpactPosition(const I3Direction &dir, I3RandomService &rng) const
{
	// first, pick which face it's going to hit
	double area = 0;
	double height = GetLength();
	std::vector<double> prob;
	std::pair<double, double> x_range(std::numeric_limits<double>::infinity(), -std::numeric_limits<double>::infinity());
	std::pair<double, double> y_range(std::numeric_limits<double>::infinity(), -std::numeric_limits<double>::infinity());
	
	BOOST_FOREACH(const polygon::side &sidey, sides_) {
		double inner = sidey.normal*dir;
		if (inner < 0)
			area += -inner*sidey.length*height;
		prob.push_back(area);
		// Find the bounding box of the cap
		if (sidey.origin.x < x_range.first)
			x_range.first = sidey.origin.x;
		if (sidey.origin.x > x_range.second)
			x_range.second = sidey.origin.x;
		if (sidey.origin.y < y_range.first)
			y_range.first = sidey.origin.y;
		if (sidey.origin.y > y_range.second)
			y_range.second = sidey.origin.y;
	}
	area += std::abs(dir.GetZ())*cap_area_;
	prob.push_back(area);
	std::vector<double>::iterator target =
	    std::lower_bound(prob.begin(), prob.end(), rng.Uniform(0, area));
	if (target == boost::prior(prob.end())) {
		// top or bottom face
		// triangulation would be more efficient here, but also more complicated
		I3Position pos(NAN,NAN,dir.GetZ() > 0 ? z_range_.first : z_range_.second);
		do {
			pos.SetX(rng.Uniform(x_range.first, x_range.second));
			pos.SetY(rng.Uniform(y_range.first, y_range.second));
		} while (!PointInHull(pos));
		return pos;
	} else {
		// side face
		std::vector<polygon::side>::const_iterator sidey =
		    sides_.begin() + std::distance(prob.begin(), target);
		double horizontal = rng.Uniform();
		double vertical = rng.Uniform();
		return I3Position(
		    sidey->origin.x + horizontal*sidey->vector.x,
		    sidey->origin.y + horizontal*sidey->vector.y,
		    z_range_.first + vertical*height
		);
	}
}

template <typename Base>
template <typename Archive>
void
ExtrudedPolygonBase<Base>::save(Archive &ar, unsigned version) const
{
	if (version > 0)
		log_fatal_stream("Version "<<version<<" is from the future");

	ar & make_nvp("Base", base_object<Base>(*this));
	std::vector<polygon::vec2> hull;
	BOOST_FOREACH(const polygon::side &sidey, sides_) {
		hull.push_back(sidey.origin);
	}
	ar & make_nvp("HullXY", hull);
	ar & make_nvp("HullZ", z_range_);
}

template <typename Base>
template <typename Archive>
void
ExtrudedPolygonBase<Base>::load(Archive &ar, unsigned version)
{
	if (version > 0)
		log_fatal_stream("Version "<<version<<" is from the future");

	ar & make_nvp("Base", base_object<Base>(*this));
	std::vector<polygon::vec2> hull;
	std::pair<double, double> z_range;
	ar & make_nvp("HullXY", hull);
	ar & make_nvp("HullZ", z_range);
	initWithHull(hull, z_range);
}

template <typename Archive>
void
ExtrudedPolygon::serialize(Archive &ar, unsigned version)
{
	if (version > 0)
		log_fatal_stream("Version "<<version<<" is from the future");

	ar & make_nvp("Base", base_object<Base>(*this));
}

}

namespace I3MuonGun {

ExtrudedPolygon::~ExtrudedPolygon() {}

template <typename Archive>
void
ExtrudedPolygon::serialize(Archive &ar, unsigned version)
{
	if (version > 0)
		log_fatal_stream("Version "<<version<<" is from the future");

	ar & make_nvp("Base", base_object<Base>(*this));
}

#if 0
bool
ExtrudedPolygon::operator==(const Surface &s) const
{
	const ExtrudedPolygon *other = dynamic_cast<const ExtrudedPolygon*>(&s);
	if (!other)
		return false;
	else {
		if (z_range_ != other->z_range_ || sides_.size() != other->sides_.size())
			return false;
		typedef std::vector<side>::const_iterator side_iter;
		for (side_iter a=sides_.begin(), b=other->sides_.begin(); a != sides_.end(); a++, b++) {
			if (a->origin.x != b->origin.x || a->origin.y != b->origin.y)
				return false;
		}
	}

	return true;
}


#endif

}

template class simclasses::ExtrudedPolygonBase<simclasses::SamplingSurface>;
template class simclasses::ExtrudedPolygonBase<I3MuonGun::SamplingSurface>;
template class I3MuonGun::detail::UprightSurface<simclasses::ExtrudedPolygonBase<I3MuonGun::SamplingSurface> >;

I3_SERIALIZABLE(simclasses::ExtrudedPolygon);
I3_SERIALIZABLE(I3MuonGun::ExtrudedPolygon);

